radixdlt/babylon-wallet-android
# [ABW-3104] Add new cloud backup system

## Description
This PR introduces a new cloud backup system (Google Drive) that replaces the [Auto Backup](https://developer.android.com/guide/topics/data/autobackup) solution. From the old backup solution we keep only the **restore functionality in order to migrate any available profile backup to the new system** when user restores wallet from backup (onboarding flow). The new flows are the following:

### 1. User creates new wallet
- user can either connect to the new cloud backup system during the onboarding flow or to skip it
- user can then enable cloud backup from Backup screen

### 2. User restores wallet from backup
User must connect to new cloud backup system in order to restore a cloud backup.
Once user is connected then the wallet will first fetch any available cloud backups from the new system, 
- if there is at least one it will be shown as available cloud backup
- if there are not available backups then the wallet will fetch any available cloud backup from the old system. And if there is one and user restores it, the wallet will migrate this profile to the new cloud backup system.

### 3. User updates an active wallet
Wallet will show a "connect to cloud screen" at launch time if:
- has backup enabled
- and there is at least one backup available (in the old cloud back up system)

---

### What is still in progress:
- Backup warnings in Account cards of the wallet screen

## How to test
During testing you can also observe the debug logs with the tag `CloudBackup`

Case 1
1. Before you run the branch, perform a backup with the old backup system
2. Uninstall the wallet and install this branch
3. Restore wallet from backup
4. Connect to the new cloud back up system
5. The backup from the old system should be available
6. Choose and proceed
7. Once in the wallet screen make some updates (e.g. new account, persona, etc..)
8. Delete the wallet 
9. Again Restore wallet from backup
10. Connect to the new cloud back up system
11. The backup from the new system should be available (check the number of accounts/personas that is correct)
12. Proceed to the wallet

Case 2
1. Install the app on two devices/emulators with the same google account 
2. With the first device create a new wallet and connect to new cloud back up system 
3. Do some updates in the profile
4. With the second device Restore Wallet from backup - connect to cloud backup system
5. The backup of the first device should be available in the second device
6. Choose the backup and proceed to the wallet screen - make some updates
7. Now back to the first device - if the wallet is still in the foreground then make an update or launch the wallet
8. After few seconds you should see the "Wallet has been transferred" screen.
9. Choose to "Clear wallet" or "transfer control back"
10. If you choose "transfer control back" then confirm that the other device will see the "Wallet has been transferred" screen if you make an update

Case 3
1. Install the app on two or more devices/emulators with the same google account, e.g 3 emulators
2. Create new wallet for each of them
3. Choose one of the device to delete wallet
4. With the same device Restore wallet from backup
5. Connect to new cloud back up system
6. You should see 3 available cloud bakups 


## Video
https://github.com/radixdlt/babylon-wallet-android/assets/118305718/b1ad5d83-5dce-4c41-8a87-5874f57cd2e0


## PR submission checklist
- [X] I have tested all the flows for the new cloud backup system
